nrf52-hal-common/src/uarte_async.rs

```diff
@@ -0,0 +1,288 @@
+
+/// Interface to a UARTE instance
+///
+/// This is a very basic interface that comes with the following limitations:
+/// - The UARTE instances share the same address space with instances of UART.
+///   You need to make sure that conflicting instances
+///   are disabled before using `Uarte`. See product specification:
+///     - nrf52832: Section 15.2
+///     - nrf52840: Section 6.1.2
+pub struct UarteAsync<T>{
+    periph: T,
+    prod: Producer,
+}
+
+impl<T> UarteAsync<T> where T: UarteAsyncExt {
+    pub fn new(
+        uarte: T,
+        nvic: &mut NVIC,
+        mut pins: Pins,
+        parity: Parity,
+        baudrate: Baudrate,
+        prod: Producer,
+        cons: Consumer,
+    ) -> Self {
+        // Select pins
+        pins.rxd.set_high();
+        uarte.psel.rxd.write(|w| {
+            let w = unsafe { w.pin().bits(pins.rxd.pin) };
+            w.connect().connected()
+        });
+        pins.txd.set_high();
+        uarte.psel.txd.write(|w| {
+            let w = unsafe { w.pin().bits(pins.txd.pin) };
+            w.connect().connected()
+        });
+
+        // Optional pins
+        uarte.psel.cts.write(|w| {
+            if let Some(ref pin) = pins.cts {
+                let w = unsafe { w.pin().bits(pin.pin) };
+                w.connect().connected()
+            } else {
+                w.connect().disconnected()
+            }
+        });
+
+        uarte.psel.rts.write(|w| {
+            if let Some(ref pin) = pins.rts {
+                let w = unsafe { w.pin().bits(pin.pin) };
+                w.connect().connected()
+            } else {
+                w.connect().disconnected()
+            }
+        });
+
+        // Enable UARTE instance
+        uarte.enable.write(|w|
+            w.enable().enabled()
+        );
+
+        // Configure
+        let hardware_flow_control = pins.rts.is_some() && pins.cts.is_some();
+        uarte.config.write(|w|
+            w.hwfc().bit(hardware_flow_control)
+             .parity().variant(parity)
+        );
+
+        // Configure frequency
+        uarte.baudrate.write(|w|
+            w.baudrate().variant(baudrate)
+        );
+
+        unsafe {
+            MAYBE_CONSUMER = Some(cons)
+        };
+
+        nvic.enable(Interrupt::UARTE0_UART0);
+
+        uarte.intenset.write(|w| {
+            w.endtx().set_bit()
+        });
+
+        UarteAsync {
+            periph: uarte,
+            prod,
+        }
+    }
+
+    pub fn write_async(
+        &mut self,
+        tx_buffer: &[u8]
+    ) -> Result<(), Error> {
+        let w_grant = self.prod.grant(tx_buffer.len())
+            .map_err(|_| Error::TxBufferTooLong)?;
+
+        w_grant.buf.copy_from_slice(tx_buffer);
+
+        self.prod.commit(tx_buffer.len(), w_grant);
+
+        NVIC::pend(Interrupt::UARTE0_UART0);
+
+        Ok(())
+    }
+
+    /// Return the raw interface to the underlying UARTE peripheral
+    pub fn free(self) -> T {
+        self.periph
```

Comment:
I should probably disable interrupts and close any open DMA transactions here, and we should maybe disable the UARTE peripheral before giving it back.